src/algebra/char_zero.lean

```diff
@@ -108,10 +108,58 @@ end
 section
 variables {R : Type*} [semiring R] [no_zero_divisors R] [char_zero R]
 
+@[simp]
 lemma add_self_eq_zero {a : R} : a + a = 0 ↔ a = 0 :=
 by simp only [(two_mul a).symm, mul_eq_zero, two_ne_zero', false_or]
 
+@[simp]
 lemma bit0_eq_zero {a : R} : bit0 a = 0 ↔ a = 0 := add_self_eq_zero
+@[simp]
+lemma zero_eq_bit0 {a : R} : 0 = bit0 a ↔ a = 0 :=
+by { rw [eq_comm], exact bit0_eq_zero }
+end
+
+section
+variables {R : Type*} [ring R] [no_zero_divisors R] [char_zero R]
+
+lemma nat_mul_inj {n : ℕ} {a b : R} (h : (n : R) * a = (n : R) * b) : n = 0 ∨ a = b :=
+begin
+  rw [←sub_eq_zero, ←mul_sub, mul_eq_zero, sub_eq_zero] at h,
+  exact_mod_cast h,
+end
+
+lemma nat_mul_inj' {n : ℕ} {a b : R} (h : (n : R) * a = (n : R) * b) (w : n ≠ 0) : a = b :=
+by simpa [w] using nat_mul_inj h
```

Comment:
Can you instead provide this instance?
```
instance [seimmodule ℕ R] : no_zero_smul_divisors ℕ R := sorry
```
Then these lemmas should fall out from `nsmul_eq_mul_cast` and `nsmul_eq_smul`.

Reply:
Neither `semimodule` nor `no_zero_smul_divisors` is available via current imports to `algebra.char_zero`, so I'm hesitant to refactor in this way.

Reply:
Hmm, that's unfortunate. I guess proving it in the other direction is just as good though; using these new lemmas to create the instance. Fine to leave that for someone else to do in a follow-up PR, but I think it would be good to have some more instances of `no_zero_smul_divisors`.